File: competition_ceginfo_2016/force_brute.py
import sys
from math import factorial

sys.setrecursionlimit(1000000)

class Tree:
    def __init__(self, val, left=None, right=None):
        self.val = val

        self.left = left
        self.right = right

        self.n_nodes = 1

        if (left is not None):
            self.n_nodes += left.n_nodes
        if (right is not None):
            self.n_nodes += right.n_nodes



def construct_tree(tree):
    if (not tree):
        return None

    i, p = 1, 0
    left = ""

    while (tree[i] != '.' or p != 0):
        if (tree[i] == '('):
            p += 1
        if (tree[i] == ')'):
            p -= 1

        left += tree[i]
        i += 1

    i += 1 # we found the root, ignore '.'

    val = ""

    while (tree[i] != '.'):
        val += tree[i]
        i += 1

    i += 1 # ignore '.'

    right = tree[i:-1]

    tree = Tree(int(val), construct_tree(left), construct_tree(right))

    return tree



def max_cost(tree):
    if (tree is None):
        return 0

    if (tree.left is None and tree.right is None):
        return 0

    return tree.val + max(max_cost(tree.left), max_cost(tree.right))



def n_binary_trees(n):
    res = [1, 1]

    for i in range(2, n+1):
        k = 0

        for j in range(1, i+1):
            k += res[j-1] * res[i-j]

        res.append(k)

    return res[n]



def print_infix_order(tree):
    if (tree is None):
        return ""

    print("(", end='')
    print_infix_order(tree.left)
    print("." + tree.val + ".", end='')
    print_infix_order(tree.right)
    print(")", end='')

    return ""

if __name__ == "__main__":
    tree = construct_tree(sys.stdin.readline().rstrip())
    print(max_cost(tree))
    print(n_binary_trees(tree.n_nodes))

File: competition_ceginfo_2016/force_brute.cpp
#include <cstddef>
#include <algorithm>
#include <iostream>
#include <gmp.h>
#include <gmpxx.h>
#include <string>

using namespace std;

struct tree
{
    tree(long val_) : val(val_), n_nodes(1), left(nullptr), right(nullptr) {}

    tree(long val_, tree* left_, tree* right_) : val(val_), left(left_), right(right_)
    {
        n_nodes = 1;

        if (left_ != nullptr)
            n_nodes += left_->n_nodes;

        if (right_ != nullptr)
            n_nodes += right_->n_nodes;
    }

    long val;
    long n_nodes;

    tree* left;
    tree* right;
};



tree* construct_tree(string t)
{
    if (t.empty())
        return nullptr;

    string left, right;

    int p = 0;

    string::const_iterator i = t.begin() + 1; // ignore the '('

    while (i != t.end())
    {
        if (*i == '(')
            p++;

        if (*i == ')')
            p--;

        if (*i == '.' && p == 0) // found the root
            break;
        else // still in the left subtree
        {
            left += *i;
            i++;
        }
    }

    i++; // ignore the '.'

    string val;

    while (*i != '.')
    {
        val += *i;
        i++;
    }

    i++; // ignore the '.'

    while (i != t.end())
    {
        right += *i;
        i++;
    }

    right.pop_back(); // ignore the last ')'

    return new tree(stol(val), construct_tree(left), construct_tree(right));
}



long max_cost(tree* t)
{
    if (t == nullptr)
        return 0;

    if (t->left == nullptr && t->right == nullptr) // leaf
        return 0;

    return (t->val + max(max_cost(t->left), max_cost(t->right)));
}



mpz_class n_binary_trees(int k)
{
    vector<mpz_class> res(k+1);

    res[0] = 1;
    res[1] = 1;

    for (int i = 2; i <= k; i++)
    {
        mpz_class n = 0;

        for (int j = 1; j <= i; j++)
            n += res[j-1] * res[i-j];

        res[i] = n;
    }

    return res[k];
}



int main()
{
   /* 
    string tree_s;
    cin >> tree_s;

    tree* t = construct_tree(tree_s);

    cout << max_cost(t) << endl;
    cout << n_binary_trees(t->n_nodes) << endl;
    */

    int i;
    cin >> i;

    cout << n_binary_trees(i) << endl;
}
